src/modm/ui/color/rgb565.hpp

```diff
@@ -0,0 +1,138 @@
+/*
+ * Copyright (c) 2021, Thomas Sommer
+ *
+ * This file is part of the modm project.
+ *
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ */
+// ----------------------------------------------------------------------------
+
+#ifndef MODM_COLOR_RGB565_HPP
+#define MODM_COLOR_RGB565_HPP
+
+#include "rgb.hpp"
+#include "hsv.hpp"
+#include "brightness.hpp"
+
+#include <modm/math/utils/clip.hpp>
+
+namespace modm
+{
+namespace color
+{
+
+// forward declarations for convertion constructors
+template<typename T>
+class RgbT;
+
+template<typename T>
+class HsvT;
+
+template<typename T>
+class BrightnessT;
+
+/**
+ * Color in RGB Colorspace, 16 bits: RRRR RGGG GGGB BBBB
+ *
+ * @author	Fabian Greif, Thomas Sommer
+ * @ingroup	modm_ui_color
+ */
+
+class Rgb565
+{
+public:
+	uint16_t color{0x0000};
+
+	using RgbCalcType = RgbT<uint8_t>;
+
+	constexpr Rgb565() = default;
+
+	/**
+	 * Constructor for preformatted Rgb565 Color
+	 *
+	 * \param color		Preformatted RGB-color in 565-format: RRRR RGGG GGGB BBBB
+	 */
+	constexpr Rgb565(uint16_t color) : color(color) {}
+
+	/**
+	 * Constructor for components: red, green, blue
+	 *
+	 * \param red 		Red color component
+	 * \param green 	Green color component
+	 * \param blue 		Blue color component
+	 */
+	constexpr Rgb565(uint8_t red, uint8_t green, uint8_t blue)
+		: color(uint16_t(red & 0xF8) << 8 | uint16_t(green & 0xFC) << 3 | uint16_t(blue >> 3))
+	{}
+
+	/**
+	 * Convertion Constructor for RGB Color
+	 *
+	 * \param rgb	RGB Color
+	 */
+	template<typename _T>
+	constexpr Rgb565(const RgbT<_T> &rgb) : Rgb565(rgb.red, rgb.green, rgb.blue)
+	{}
+
+	/**
+	 * Convertion Constructor for HSV Color
+	 *
+	 * \param hsv	HSV Color
+	 */
+	template<typename _T>
+	constexpr Rgb565(const HsvT<_T> &hsv) : Rgb565(RgbCalcType(hsv))
+	{}
+
+	/**
+	 * Convertion Constructor for Brightness
+	 *
+	 * \param brightness	Brightness 'Color'-object
+	 */
+	template<typename _T>
+	constexpr Rgb565(const BrightnessT<_T> brightness) : Rgb565(RgbCalcType(brightness))
+	{}
+
+	/**
+	 * Convertion function to plain color
+	 */
+	inline operator uint16_t() const {
```

Comment:
Do we really need an implicit conversion here? This would make the following code legal:
```
modm::color::Rgb565 rgb = ...;
rgb = 2.3f * rgb;
```
It wouldn't do the right thing but compiles. What happens is that at first `rgb` is converted to `uint16_t`, that is converted to `float`, multiplied, narrowed to `uint16_t` and implicitly converted to `uint16_t` again.

Reply:
Maybe this was a drip too much of implicit comfort 🙄?